proposals/CodifiedVUs.adoc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 The Khronos Group Inc.
+//
+// SPDX-License-Identifier: CC-BY-4.0
+
+= Codified VUs
+:toc: left
+:refpage: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/
+:sectnums:
+
+In this proposal, a new procedural syntax for writing Valid Usage (VU)
+statements in the spec is introduced.
+
+== Problem Statement
+
+Originally, Vulkan VUs were written in prose (and continue to do so).
+This creates a number of difficulties.  Take the following VU for example:
+
+[source,asciidoc]
+~~~~
+  * [[VUID-VkRenderingInfo-colorAttachmentCount-06097]]
+    If pname:colorAttachmentCount is not `0` and the pname:imageView member
+    of an element of pname:pColorAttachments is not dlink:VK_NULL_HANDLE, if
+    the pname:resolveMode member of that element of pname:pColorAttachments
+    is not ename:VK_RESOLVE_MODE_NONE, its pname:resolveImageLayout member
+    must: not be
+    ename:VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_STENCIL_ATTACHMENT_OPTIMAL or
+    ename:VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_STENCIL_READ_ONLY_OPTIMAL
+~~~~
+
+Written in this form, there are typically long descriptions of otherwise simple
+constructs, and back references such as "that element" or "its".
+The result is hard to interpret and sometimes ambiguous VUs.
+
+In essence, these VUs are basic algorithms composed of loops and conditions
+leading to assertions.
+However, a natural language (such as English) is not best suited to describe this.
+The authors of the spec, its users and the implementers of the Vulkan
+Validation Layers (VVL) are typically all programmers, which makes use of prose
+as the medium of communicating assertions even more awkward.
+
+== Proposal
+
+We could do better by writing VUs in a programming language instead.
+In the above example, one can identify a for loop, two conditions and an
+assertion.
+That could be written as:
+
+[source,python]
+~~~~
+  * [[VUID-VkRenderingInfo-colorAttachmentCount-06097]]
+    for attachment in pColorAttachments:
+     if (attachment.imageView != VK_NULL_HANDLE and
+         attachment.resolveMode != VK_RESOLVE_MODE_NONE):
+      resolveLayout = attachment.resolveImageLayout
+      require(resolveLayout != VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_STENCIL_ATTACHMENT_OPTIMAL and
```

Comment:
One readability thing we're likely to run into is that many Vulkan tokens are extremely long, and come near to running off the right side of the window as is, so will be much more common to see that when they are formatted as source code.

Reply:
Yes indeed that's partially why this example uses a variable to make the LHS of the comparison shorter. I would very much like to fix that, especially because line wraps and python scoping don't really mix well.

In fact, we already have this problem with our promoted enums (see wrapped lines at the bottom of these enums for example): https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkFormatFeatureFlagBits.html

IMO, the best way to deal with that is to have code blocks (not just these VUs, but also struct definitions, enums, etc) be under a div that can get horizontal scrolling. We could generally avoid that by structuring our sources to use shorter names for long expressions, but when not possible, it'll be good to have that automatically avoid overflow.

Another way to (mostly) address it by the way is to enhance the formatter to track length of expressions and automatically break expressions into multiple lines. We have these sort of code formatters in Chromium scripts, but I think a generic one like that is going to be too much complexity. Right now, what I have coded is unconditional line breaks between "and" and "or" expressions, and no line breaks otherwise. I'm hoping that would be generally sufficient for VUs.

Reply:
FWIW, currently in #2053 I've opted for disabling line wrapping for codified VUs.